File: frontend/src/lib/api/workspace.ts
import { pythonFetch } from "./client";

export interface Workspace {
  id: string;
  name: string;
  slug: string;
  description?: string;
  owner_id?: string;
  created_at: string;
  organization_id: string;
}

export async function fetchWorkspaces(): Promise<Workspace[]> {
  const response = await pythonFetch("workspaces");
  if (!response.ok) {
    throw new Error("Failed to fetch workspaces");
  }
  return response.json();
}

export async function fetchWorkspaceDetails(workspaceId: string): Promise<Workspace> {
  const response = await pythonFetch(`workspaces/${workspaceId}`);
  if (!response.ok) {
    throw new Error("Failed to fetch workspace details");
  }
  return response.json();
}

export async function provisionWorkspace(data: {
  organization_name: string;
  organization_slug: string;
  workspace_name: string;
  workspace_slug: string;
}): Promise<Workspace> {
  const response = await pythonFetch("workspaces/provision", {
    method: "POST",
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Failed to provision workspace");
  }
  return response.json();
}

export async function updateWorkspace(workspaceId: string, data: {
  name?: string;
  description?: string;
}): Promise<Workspace> {
  const response = await pythonFetch(`workspaces/${workspaceId}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Failed to update workspace");
  }
  return response.json();
}

export interface Invitation {
  id: string;
  email: string;
  workspace_id: string;
  role: string;
  invited_by: string;
  token: string;
  status: string;
  expires_at: string;
}

export async function inviteMember(workspaceId: string, email: string, role: string = "member"): Promise<Invitation> {
  const response = await pythonFetch(`workspaces/${workspaceId}/invitations`, {
    method: "POST",
    body: JSON.stringify({ email, role }),
  });
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Failed to invite member");
  }
  return response.json();
}

File: python-service/app/schemas/workspace.py
import uuid
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, EmailStr

class OrganizationBase(BaseModel):
    name: str
    slug: str

class OrganizationCreate(OrganizationBase):
    pass

class OrganizationResponse(OrganizationBase):
    id: uuid.UUID
    
    model_config = ConfigDict(from_attributes=True)

class WorkspaceBase(BaseModel):
    name: str
    slug: str

class WorkspaceCreate(WorkspaceBase):
    organization_id: uuid.UUID

class WorkspaceResponse(WorkspaceBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    description: Optional[str] = None
    owner_id: Optional[uuid.UUID] = None
    created_at: datetime
    
    model_config = ConfigDict(from_attributes=True)

class WorkspaceProvision(BaseModel):
    organization_name: str
    organization_slug: str
    workspace_name: str
    workspace_slug: str

class InvitationCreate(BaseModel):
    email: EmailStr
    role: str = "member"

class InvitationResponse(BaseModel):
    id: uuid.UUID
    email: EmailStr
    workspace_id: uuid.UUID
    role: str
    invited_by: uuid.UUID
    token: str
    status: str
    expires_at: datetime
    
    model_config = ConfigDict(from_attributes=True)

class InvitationAccept(BaseModel):
    token: str

